source/slang/slang-emit-c-like.cpp

```diff
@@ -2169,13 +2224,41 @@ void CLikeSourceEmitter::defaultEmitInstExpr(IRInst* inst, const EmitOpInfo& inO
         }
         else
         {
-            auto prec = getInfo(EmitOp::Postfix);
-            needClose = maybeEmitParens(outerPrec, prec);
+            if (inst->op == kIROp_getElementPtr && supportPtrTypes())
+            {
+                auto ptrType = as<IRPtrTypeBase>(inst->getOperand(0)->getFullType());
+                if (ptrType && ptrType->getValueType() && as<IRArrayTypeBase>(ptrType->getValueType()))
+                {
+                    const auto info = getInfo(EmitOp::Postfix);
+                    needClose = maybeEmitParens(outerPrec, info);
+                    emitOperand(inst->getOperand(0), leftSide(outerPrec, info));
+                    m_writer->emit("->elementPtr(");
+                    emitOperand(inst->getOperand(1), getInfo(EmitOp::General));
+                    m_writer->emit(")");
+                }
+                else
+                {
+                    const auto info = getInfo(EmitOp::Add);
+
+                    needClose = maybeEmitParens(outerPrec, info);
+                    emitOperand(inst->getOperand(0), leftSide(outerPrec, info));
+                    m_writer->emit(" ");
+                    m_writer->emit(info.op);
+                    m_writer->emit(" ");
+                    emitOperand(inst->getOperand(1), rightSide(outerPrec, info));
+                }
+                break;
+            }
+            else
+            {
+                auto prec = getInfo(EmitOp::Postfix);
+                needClose = maybeEmitParens(outerPrec, prec);
 
-            emitOperand( inst->getOperand(0), leftSide(outerPrec, prec));
-            m_writer->emit("[");
-            emitOperand(inst->getOperand(1), getInfo(EmitOp::General));
-            m_writer->emit("]");
+                emitOperand(inst->getOperand(0), leftSide(outerPrec, prec));
+                m_writer->emit("[");
+                emitOperand(inst->getOperand(1), getInfo(EmitOp::General));
+                m_writer->emit("]");
+            }
```

Comment:
If we have an operation like `emitDereferenceOperand` that handles emitting a `&` if needed, and skipping it if not, we could collapse all these cases down to one.

The main difficulty with an operation that emits prefix `&` or `*` as needed is that it might lead to parentheses being needed. It seems like overall the `needClose` setup in this code should change from `bool needClose` to `int closeCount` which tracks the number of closing `)` needed. Then we can have steps that do `closeCount += maybeEmitPrefixAddrOf(...)`.

In the far-flung future, a major alternative here would be to have the emit logic build up a "rope" structure in a bottom-up fashion, rather than the current approach that emits text directly in a top-down fashion.

Reply:
Change of `needClose` is a big change on its own, so I'd rather leave that as a future improvement if we determine it is necessary.